aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1364,81 +1364,515 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+      auto* ptr_cidx = cidx.data_ptr<int64_t>();
+
+      const auto idx_len = idx.numel();
+
+      std::fill_n(ptr_cidx, ptr_idx[0] + 1, 0);
+      std::fill_n(ptr_cidx + ptr_idx[idx_len - 1] + 1, len - ptr_idx[idx_len - 1], idx_len);
+
+      const auto grain_size = run_in_parallel ? at::internal::GRAIN_SIZE : idx_len;
+      at::parallel_for(0, idx_len, grain_size, [&](int64_t start, int64_t end) {
+          auto* ptr_curr_cidx = ptr_cidx + ptr_idx[start] + 1;
+          for (int64_t i = start; i < std::min(end, idx_len - 1); ++i) {
+            const auto diff = ptr_idx[i + 1] - ptr_idx[i];
+            std::fill_n(ptr_curr_cidx, diff, i + 1);
+            ptr_curr_cidx += diff;
+          }
+      });
+
+      return cidx;
+    };
+
+    // If nnz is (much) larger than size, then both indices[dim] and index get sorted
+    // with a count sort (faster, and no huge nnz-sized chunk memory allocations).
+    // The element-wise product between the count tables gives us all the intersections.
+    const auto get_selected_indices_large_nnz_small_size = [&]() -> std::tuple<Tensor, Tensor> {
+      const auto get_counts = [&sorted_idx_to_cidx](
+          // Writes into counts (must be preallocated and zero)
+          // and allows to use external buffers.
+          Tensor& counts,
+          const Tensor& t,
+          int64_t bins,
+          bool is_sorted = false,
+          bool run_in_parallel = true) -> void {
+        if (is_sorted) {
+          const auto cidx = sorted_idx_to_cidx(t, bins, run_in_parallel);
+          at::sub_out(counts, cidx.slice(0, 1, bins + 1), cidx.slice(0, 0, bins));
+        }
+        else {
+          auto* ptr_counts = counts.data_ptr<int64_t>();
+          const auto* ptr_vals = t.data_ptr<int64_t>();
+          for (C10_UNUSED const auto _ : c10::irange(t.numel())) {
+            ++ptr_counts[*ptr_vals++];
+          }
         }
+      };
+
+      const auto counts_per_thread = [&get_counts, size](
+          const Tensor& idx,
+          bool is_sorted = false,
+          int64_t grain_size = at::internal::GRAIN_SIZE
+      ) -> Tensor {
+        const auto idx_len = idx.numel();
+        // 1 <= n_threads <= min(ceil(len / grain_size), max_threads)
+        const auto n_threads = std::max<int64_t>(
+            1, std::min<int64_t>((idx_len + grain_size - 1) / grain_size, at::get_num_threads())
+        );
+
+        auto counts_per_thread = at::zeros({n_threads, size}, idx.options());
+        at::parallel_for(0, idx_len, grain_size, [&](int64_t start, int64_t end) {
+          const auto tid = at::get_thread_num();
+          const auto tid_idx = idx.slice(0, start, end);
+          auto tid_counts = counts_per_thread.select(0, tid);
+          get_counts(tid_counts, tid_idx, /*bins=*/size,
+              /*is_sorted=*/is_sorted, /*run_in_parallel=*/false);
+        });
+
+        return counts_per_thread;
+      };
+
+      const bool search_in_dim_indices = true;
+
```

Comment:
Left here intentionally. I still have to think about a better heuristic. Also applies to the case `nnz > size`, and the number of threads should be taken into consideration.